storage/snapshot/snapshot.rs

```diff
@@ -94,17 +94,20 @@ pub trait WritableSnapshot: ReadableSnapshot {
     ) -> Result<ByteArray<BUFFER_VALUE_INLINE>, SnapshotGetError> {
         let keyspace_id = key.keyspace_id();
         let writes = self.operations().writes_in(keyspace_id);
-        let existing = writes.get(key.bytes());
-        if let Some(existing) = existing {
-            Ok(existing)
+        if let Some(operation) = writes.get(key.bytes()) {
+            match operation {
+                Write::Insert {value, .. } | Write::Put {value, .. } => {
+                    Ok(ByteArray::copy(value.bytes()))
+                },
+                Write::Delete => Err(SnapshotGetError::GetRequiredKeyDoesntExist { key: StorageKey::Array(key.into_owned_array()) })
+            }
         } else {
             let storage_value = self.get_mapped(key.as_reference(), |reference| ByteArray::from(reference))?;
             if let Some(value) = storage_value {
                 self.operations_mut().lock_add(ByteArray::copy(key.bytes()), LockType::Unmodifiable);
                 Ok(value)
             } else {
-                // TODO: what if the user concurrent requires a concept while deleting it in another query
-                unreachable!("Require key exists in snapshot or in storage.");
+                Err(SnapshotGetError::GetRequiredKeyDoesntExist { key: StorageKey::Array(key.into_owned_array()) })
```

Comment:
Is this really an illegal state worthy of panic? Or should we be returning an error?

Reply:
From the global perspective, yes.
But from the perspective of a snapshot?
Should we panic here or should the callers call unwrap?